cpp/oneapi/dal/backend/transfer_dpc.cpp

```diff
@@ -98,33 +103,73 @@ sycl::event scatter_host2device(sycl::queue& q,
                                       block_count * block_size_in_bytes,
                                       deps);
 
-    auto scatter_event = q.submit([&](sycl::handler& cgh) {
-        cgh.depends_on(copy_event);
-
-        byte_t* gathered_byte = reinterpret_cast<byte_t*>(gathered_device_unique.get());
-        byte_t* dst_byte = reinterpret_cast<byte_t*>(dst_device);
-
-        const std::int64_t required_local_size = 256;
-        const std::int64_t local_size = std::min(down_pow2(block_count), required_local_size);
-        const auto range = make_multiple_nd_range_1d(block_count, local_size);
-
-        cgh.parallel_for(range, [=](sycl::nd_item<1> id) {
-            const auto i = id.get_global_id();
-            if (i < block_count) {
-                // TODO: Unroll for optimization
-                for (int j = 0; j < block_size_in_bytes; j++) {
-                    dst_byte[i * dst_stride_in_bytes + j] =
-                        gathered_byte[i * block_size_in_bytes + j];
+    byte_t* gathered_byte = reinterpret_cast<byte_t*>(gathered_device_unique.get());
+    byte_t* dst_byte = reinterpret_cast<byte_t*>(dst_device);
+
+    const std::int64_t required_local_size = 256;
+
+    const std::int64_t max_loop_range = std::numeric_limits<std::int32_t>::max();
+
+    if (block_count > max_loop_range) {
+        const auto block_size = propose_block_size<float>(q, block_count);
+        const bk::uniform_blocking blocking(block_count, block_size);
+        std::vector<sycl::event> events(blocking.get_block_count());
+
+        for (std::int64_t block_index = 0; block_index < blocking.get_block_count();
+             ++block_index) {
+            const auto start_block = blocking.get_block_start_index(block_index);
+            const auto end_block = blocking.get_block_end_index(block_index);
+            const auto curr_block = end_block - start_block;
+            ONEDAL_ASSERT(curr_block > 0);
+
+            auto scatter_event = q.submit([&](sycl::handler& cgh) {
+                cgh.depends_on(copy_event);
+
+                const std::int64_t required_local_size = 256;
```

Comment:
There is a second definition, `required_block_size` is defined above. It could probably be removed.
Is 256 always optimal? Is `bk::device_max_wg_size(queue)` not optimal?